### ISSUE TITLE
Error in Microsoft.ML.Extensions (Attempted to divide by zero.) when running in Azure

### ISSUE THREAD

**Issue by jankozeleny** [dotnet/machinelearning#4171]
### System information

- Azure AppService
- .NET Core 2.2, 64-bit

### Issue

Deploy the unmodified E2E sample https://github.com/dotnet/machinelearning-samples/tree/v1.2/samples/csharp/end-to-end-apps/DeepLearning_ObjectDetection_Onnx
into Azure AppService.
Note that you must switch to 64-bit because the AppService will not even start (I suppose that this is a problem with Core 2.2 in-process hosting but I am not sure).
Also you must modifiy the relative path in the Get action in the ObjectDetectionController
`string imageFileRelativePath = @"assets" + url;`

**The code stops at the unhandled division by zero exception** ("Attempted to divide by zero.") at line:
`            var probs = model.Predict(imageInputData).PredictedLabels;`
in the ObjectDetectionService in the DetectObjectsUsingModel method.

**The application runs without problem when running locally.**

Stack trace:

```
   at Microsoft.ML.OnnxRuntime.NativeMethods.OrtRun(IntPtr session, IntPtr runOptions, String[] inputNames, IntPtr[] inputValues, UInt64 inputCount, String[] outputNames, UInt64 outputCount, IntPtr[] outputValues)
   at Microsoft.ML.OnnxRuntime.InferenceSession.Run(IReadOnlyCollection`1 inputs, IReadOnlyCollection`1 outputNames, RunOptions options)
   at Microsoft.ML.OnnxRuntime.InferenceSession.Run(IReadOnlyCollection`1 inputs, IReadOnlyCollection`1 outputNames)
   at Microsoft.ML.OnnxRuntime.InferenceSession.Run(IReadOnlyCollection`1 inputs)
   at Microsoft.ML.Transforms.Onnx.OnnxTransformer.Mapper.UpdateCacheIfNeeded(Int64 position, INamedOnnxValueGetter[] srcNamedOnnxValueGetters, String[] activeOutputColNames, OnnxRuntimeOutputCacher outputCache)
   at Microsoft.ML.Transforms.Onnx.OnnxTransformer.Mapper.<>c__DisplayClass11_0`1.<MakeTensorGetter>b__0(VBuffer`1& dst)
   at Microsoft.ML.Data.TypedCursorable`1.TypedRowBase.<>c__DisplayClass8_0`1.<CreateDirectVBufferSetter>b__0(TRow row)
   at Microsoft.ML.Data.TypedCursorable`1.TypedRowBase.FillValues(TRow row)
   at Microsoft.ML.PredictionEngineBase`2.Predict(TSrc example)
   at Microsoft.Extensions.ML.PredictionEnginePoolExtensions.Predict[TData,TPrediction](PredictionEnginePool`2 predictionEnginePool, String modelName, TData example)
   at OnnxObjectDetectionE2EAPP.Services.ObjectDetectionService.DetectObjectsUsingModel(ImageInputData imageInputData) in C:\p\test\DeepLearning_ObjectDetection_Onnx\OnnxObjectDetectionE2EAPP\Services\ObjectDetectionService.cs:line 29
   at OnnxObjectDetectionE2EAPP.Controllers.ObjectDetectionController.DetectAndPaintImage(ImageInputData imageInputData, String imageFilePath) in C:\p\test\DeepLearning_ObjectDetection_Onnx\OnnxObjectDetectionE2EAPP\Controllers\ObjectDetectionController.cs:line 125
   at OnnxObjectDetectionE2EAPP.Controllers.ObjectDetectionController.Get(String url) in C:\p\test\DeepLearning_ObjectDetection_Onnx\OnnxObjectDetectionE2EAPP\Controllers\ObjectDetectionController.cs:line 59
```

I am now clueless and I even don't know how .NET DivideByZeroException can be raised in the C interop call. Where does it come from?

**Comment by jankozeleny** [comment]
I forgot HResult from the exception is -2147352558.

**Comment by mdslack** [comment]
I get the same divide by zero exception at NativeMethods.OrtRun when I run the ONNX Mask-R-CNN object detection model using the GPU ONNX runtime. The same code works perfectly on the CPU version of ONNX runtime.

**Comment by luisquintanilla** [comment]
Thanks for reporting this @jankozeleny and @mdslack . I noticed that the referenced sample is using v1.2. Do you get the same issues when working with v 1.3.1?

This is the link to the most up to date sample. The `OnnxObjectDetectionWeb` project is the one to look at since it shares the functionality of the sample you referenced in this issue. 

https://github.com/dotnet/machinelearning-samples/tree/master/samples/csharp/end-to-end-apps/ObjectDetection-Onnx

**Comment by eerhardt** [comment]
@jignparm @shahasad - any thoughts here? The error appears to be coming from the OnnxRuntime.

@jankozeleny and @mdslack - maybe opening an issue on https://github.com/microsoft/onnxruntime/, especially if this repros outside of ML.NET just using the OnnxRuntime API.

**Comment by mdslack** [comment]
> Thanks for reporting this @jankozeleny and @mdslack . I noticed that the referenced sample is using v1.2. Do you get the same issues when working with v 1.3.1?

Sorry, I didn't mention it in my previous message, but I'm actually running v 1.4.0-preview. I need this version for my project as it has the update needed for using the ONNX Mask-R-CNN model from the model zoo [#3963](https://github.com/dotnet/machinelearning/pull/3963).

**Comment by asturm0** [comment]
Any update on this? I get the same error with v1.4 on Azure App Service hosting a NET Core 2.2 app.

**Comment by mareklinka** [comment]
I just ran into this very frustrating issue as well.

.NET Core SDK 3.1.102
Microsoft.ML (+ImageAnalytics, OnnxTransformer) 1.4.0

When trying to deploy a very simple app that just takes a pretrained ONNX model and exposes it as an API, all requests break down with "Attempted to divide by zero" in a Windows app service.

Everything works normally on local, of course.

In the end, I came up with a workaround, if anyone comes across this: instead of a Windows app service, I deployed the app into a Linux app service. That lead to a few problems because some of the ONNX-required libraries were not present in the Linux docker image (Linux app services run as containers). Namely, libgomp1 and libgdiplus (probably because the model is a ConvNet). Both of these can be installed with apt-get.

Since the container might get restarted (causing local changes to be lost), I had to get creative with my entrypoint. I created a run.sh script like this:

``` bash
apt-get update
apt-get install -y libgomp1
apt-get install -y libgdiplus

Project.executable
```
And then I pointed Azure at this script:
![image](https://user-images.githubusercontent.com/7479226/76697518-3b76de80-66d3-11ea-8adc-3f760868ad61.png)

Now each time the container starts, it installs its dependencies and the application can run without issues. It's far from an optimal solution (the startup time suffers considerably), but it works. Still, I hope the division by zero gets fixed soon.

**Comment by eerhardt** [comment]
@mareklinka - Can you try the latest release [`1.5.0-preview2`](https://www.nuget.org/packages/Microsoft.ML.OnnxTransformer/1.5.0-preview2)? I believe this issue was fixed in the latest release because ML.NET integrated a new version of the OnnxRuntime (where this bug existed).

**Comment by mareklinka** [comment]
Damn, I literally spent hours on this yesterday when I could have solved in in 5 minutes 😄 

Update to 1.5.0-preview2 solved the problem. There is one caveat, though, not sure if it's expected so I'll just mention it.

My original setup was using the following nuget packages (all 1.4.0):
```
Microsoft.Extensions.ML
Microsoft.ML
Microsoft.ML.ImageAnalytics
Microsoft.ML.OnnxTransformer
```
And this worked out of the box locally. When I updated to 1.5.0-preview2, even local debugging stopped working because the native onnxruntime could not be found. I checked the dependency tree and it looks like this:
```
Microsoft.ML.OnnxTransformer
    Microsoft.ML.OnnxRuntime.Managed
```
However, the `Microsoft.ML.OnnxRuntime.Managed` package does not depend on `Microsoft.ML.OnnxRuntime`. It's the other way around:
```
Microsoft.ML.OnnxRuntime
    Microsoft.ML.OnnxRuntime.Managed
```
I guess that's why the native onnx libs were not present. Anyway, after manually installing the `Microsoft.ML.OnnxRuntime` package, everything works fine locally AND in a Windows app service.

Thanks for pointing me to this!

**Comment by eerhardt** [comment]
Yes, this is expected. This allows the application to pick whether it wants CPU or GPU enabled processing. If you want the ONNX model to run on a GPU, instead of referencing `Microsoft.ML.OnnxRuntime`, you would reference https://www.nuget.org/packages/Microsoft.ML.OnnxRuntime.Gpu.

See https://github.com/microsoft/onnxruntime/issues/2184 for more information.

**Comment by mareklinka** [comment]
@eerhardt Ah, makes sense. Thanks for the clarification!

**Comment by mstfbl** [comment]
As @eerhardt mentioned above, this issue is fixed in the latest version of `OnnxRuntime` v1.2.0, which ML.NET's latest release `1.5.0-preview2` integrated.
